### ISSUE TITLE
cudnn Dropout reproducibility 

### ISSUE THREAD

**Issue by eric-haibin-lin** [apache/incubator-mxnet#15662]
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h#L491-L493

The seed of CUDNN dropout does not respect the random seed in mxnet

**Comment by zachgk** [comment]
I did a quick recursive grep and found several other usages of the rand function. Here is the full list I found (some of these may be correct):
- https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h#L95
- https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h#L491
- https://github.com/apache/incubator-mxnet/blob/master/src/operator/random/sampler.h#L110
- https://github.com/apache/incubator-mxnet/blob/master/src/operator/rnn-inl.h#L1551
- https://github.com/apache/incubator-mxnet/blob/master/src/operator/rnn_impl.h#L159
- https://github.com/apache/incubator-mxnet/blob/master/src/operator/rnn_impl.h#L1004
- https://github.com/apache/incubator-mxnet/blob/master/src/operator/rnn_impl.h#L1891

**Comment by eric-haibin-lin** [comment]
@ptrendx @DickJC123 just FYI

**Comment by larroy** [comment]
So all those places should be seeded propperly correct?

**Comment by eric-haibin-lin** [comment]
random/sampler.* and rnn_impl.* are for other operators, not dropout. But yes, in general those should also be fixed, and deserve separate github issues

**Comment by eric-haibin-lin** [comment]
below is a test case: 
```
import mxnet as mx

a = mx.nd.ones((10,10), ctx=mx.gpu())
dropout = mx.gluon.nn.Dropout(0.5)
seed = 1234

mx.random.seed(seed)

with mx.autograd.record():
    b = dropout(a)

mx.random.seed(seed)

with mx.autograd.record():
    c = dropout(a)

print(b)
print(c)
mx.test_utils.assert_almost_equal(b.asnumpy(), c.asnumpy())
```

**Comment by eric-haibin-lin** [comment]
we need to delay the initialization of the seed in https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h#L497
request a random resource: https://github.com/apache/incubator-mxnet/blob/e260f13499a6c84737a97aae09e5921ad8fdd4fd/src/operator/numpy/random/np_choice_op.cc#L71
get a random number generator: https://github.com/apache/incubator-mxnet/blob/e260f13499a6c84737a97aae09e5921ad8fdd4fd/src/operator/numpy/random/np_choice_op.h#L166
do sampling with the generator: https://github.com/apache/incubator-mxnet/blob/e260f13499a6c84737a97aae09e5921ad8fdd4fd/src/operator/numpy/random/np_choice_op.h#L179

**Comment by leezu** [comment]
Reopen as fix is reverted due to performance issue.
